ludwig/utils/horovod_utils.py

```diff
@@ -54,33 +54,51 @@ def gather_all_tensors(result: torch.Tensor, group: Optional[Any] = None) -> Lis
     :return: list with size equal to the process group where gathered_result[i]
              corresponds to result tensor from process i
     """
+    print("ASDFASDF inside gather_all_tensors")
     if group is not None:
         raise ValueError("Horovod does not support allgather using a subcommunicator at this time. " "Unset `group`.")
 
+    print("\tgroup is None")
+    print("\tresult passed in:", result)
+
     if _HVD is None or not _HVD.is_initialized():
         return [result]
 
     if len(result.shape) == 0:
         # Convert scalars to single dimension tensors
         result = result.reshape(1)
+        print("\tresult after reshape:", result)
 
     is_bool = False
     if result.dtype == torch.bool:
         # need to convert to int due to Horovod limitation
         result = result.int()
         is_bool = True
+        print("\tresult after int:", result)
 
     # Add extra dimension to the tensors to be gathered
     result = result.unsqueeze(0)
 
+    print("\tresult after unsqueeze:", result)
+
     # sync and gather all
-    gathered = _HVD.allgather(result)
-    gathered_result = list(gathered.split(1, dim=0))
+    gathered_result = _HVD.allgather(result)
+
+    print("\tgathered_result:", gathered_result)
+
+    # This is to match the output of the torchmetrics gather_all_tensors function
+    # and ensures that the return value is usable by torchmetrics.compute downstream.
+    if len(result.shape) >= 2:
```

Comment:
Interesting, so my understanding is that DDP allgather contract is like:

```
tensor[*shape] -> [tensor[*shape]] * world_size
```

While Horovod is:

```
tensor[shape] -> tensor[world_size, *shape]
```

In the previous implementation, we split the tensor along the `rank` dimension (dim=0) into `[tensor[shape]] * world_size` to match the DDP format. But with this change it seems if the input is a scalar, meaning the allgathered output is a 1D vector, then we just return it as-is (unless it's a bool, in which case we iterate over it and turn it into a list of bool tensors).

Seems like the output format is potentially inconsistent then, right? It could be a list of tensors or it could be a single tensor. Am I missing something.

The other aspect here is the difference between casting to a list vs calling split. Running this in a terminal, I get:

```
>>> t
tensor([[1., 1., 2.],
        [2., 3., 1.],
        [5., 2., 4.]])
>>> list(t)
[tensor([1., 1., 2.]), tensor([2., 3., 1.]), tensor([5., 2., 4.])]
>>> t.split(1, dim=0)
(tensor([[1., 1., 2.]]), tensor([[2., 3., 1.]]), tensor([[5., 2., 4.]]))
```

So it seems like the difference here is that `split` preserves the rank of the tensor, while `list` removes the rank that's being split upon entirely. So that does seem to better align with the DDP format.

As such, is the right thing to do here just to change lines 77 and 78 to:

```
gathered = _HVD.allgather(result)
gathered_result = list(gathered)
```

In other words, why only do this when `len(result.shape) >= 2`?

Otherwise, it makes sense, as I do believe DDP allgather does not add an extra dimension to the tensors being gathered, as the code seemed to be doing previously.

Reply:
Oh interesting– so double-checking this, it seems like the guard I've added is pretty much always triggered. I believe this is because of these two lines:

https://github.com/ludwig-ai/ludwig/blob/master/ludwig/utils/horovod_utils.py#L63-L66

https://github.com/ludwig-ai/ludwig/blob/master/ludwig/utils/horovod_utils.py#L73-L75

Tensors always have a rank of at least one because of the first code block, and the second code block ensures that the rank is at least two. I can remove the guard, but FYI I don't think it will make too much of a difference.